### ISSUE TITLE
Thread state through nom combinators

### ISSUE THREAD

**Issue by joshrule** [Geal/nom#1419]
## Prerequisites

Here are a few things you should provide to help me understand the issue:

- Rust version : `rustc 1.55.0 (c8dfcfe04 2021-09-06)`
- nom version : `7.0.0`
- nom compilation features used: None

## Test case

I'm updating [a parser](https://github.com/joshrule/term-rewriting-rs/blob/develop/src/parser.rs) from nom 4 to nom 7. I'm parsing term rewriting systems, which involves parsing a lot of first-order terms. These first-order terms can contain function symbols and constants, collectively called operators, as well as variables. The identity of each operator and variable is maintained by assigning it a numeric id, and these ids are established while parsing. The parser maintains a map between strings and ids.

In nom 4, I created a parser struct and used combinations of the method! and call_m! macros to build up the id map. In nom 7, these macros no longer exist, nor do corresponding functions.

My first attempt was to translate something like:

```rust
method!(application<Parser<'a>, CompleteStr, Term>, mut self,
        alt!(call_m!(self.standard_application) |
             call_m!(self.binary_application))
);
```

into

```rust
fn application(&mut self, s: &'a str) -> IResult<&'a str, Term> {
    alt((
        |x: &'a str| self.standard_application(x),
        |x: &'a str| self.binary_application(x),
    ))(s)
}
```

but the compiler helpfully reminds me that this requires simultaneous unique access to self in both closures.

Perhaps nom could provide the option to fold some sort of mutable state type through the combinators so that I could write:

```rust
fn application(&mut self, s: &'a str) -> IResult<&'a str, Term> {
    alt_with_state(self, (
        |x: &'a str, this /*: &mut Self */| this.standard_application(x),
        |x: &'a str, this| this.binary_application(x),
    ))(s)
}
```

**Comment by joshrule** [comment]
Additional [discussion](https://users.rust-lang.org/t/parsers-carrying-state-in-nom-7/65291) on the Rust Discourse.

**Comment by Stargateur** [comment]
I'm think you are doing it wrong, a parser should not require an external state like that. That doesn't make sense that if a parser fail it mutate the state. You would mutate the state in the top level parser in function of what your subparser return.

I'm opposite to this feature, FnMut already allow state and sharing state between parser in alt is bad design.

**Comment by joshrule** [comment]
Thanks for responding. I'm not sure exactly how to apply your advice. I'm trying to parse text directly into first-order terms whose symbols are identified with integers. Mutating state as I go along to build a map between symbol names and integer IDs seems like a natural approach that avoids multiple passes through the data. How else would you suggest I solve this problem?

**Comment by Stargateur** [comment]
> Thanks for responding. I'm not sure exactly how to apply your advice. I'm trying to parse text directly into first-order terms whose symbols are identified with integers. Mutating state as I go along to build a map between symbol names and integer IDs seems like a natural approach that avoids multiple passes through the data. How else would you suggest I solve this problem?

Unfortunately, I didn't understand much, `first-order terms` doesn't mean much without context. If possible could you make me a little example of how you would use `alt_with_state` with your parser ? I believe I already explain a solution, your parser should return result and you should use this result to build your state but maybe with a code example it would be more clear for both of us. I try to look at the code of your crate but that was too hard to follow without a lot of time to read.

**Comment by przygienda** [comment]
there is another very good use case for at least to carry read only data through the parser. Variants. Sometimes complex parsers have "well, accept this but not that for compatibility reasons, leads to bunch of alts driven by an if possibly". Completely stateless parsers are cute, like perfect FSMs. practically speaking, most realistic complex code does extended FSM with state or parsers that are nob'ed for behavior

**Comment by przygienda** [comment]
a possible solution is to actually implement the whole Input trait zoo for something that is (configuration, &[u8]) e.g. which leads quickly into a nice lifetime hell ;-) But in a sense this is the clean solution to parametrize the parser.

**Comment by epage** [comment]
For state
- `nom_locate` has an [`extra` field](https://docs.rs/nom_locate/4.2.0/nom_locate/struct.LocatedSpan.html#structfield.extra)
- `pori` has the [`Stateful` input decorator](https://docs.rs/pori/latest/pori/struct.Stateful.html)

**Comment by przygienda** [comment]
unfortunately 0.0 crate and zero doc.

Having had a quick look I basically implemented the pori Stateful stuff
myself now
